Refactor ruby interface

fables-tales opened the pull request:
<!--
Hi there! Thanks for taking the time to file  a pull request against Rubyfmt
Right now we're accepting CLI ergonomics PRs, Editor Integration PRs, and bug
fixes only. We define bugs as Rubyfmt failing to format a file, or formatting a
file such that it's behaviour changes. If you're trying to change the behaviour
of the formatter, or style the output, please don't file the PR. We're working
on getting that just right, and will accept those in a future release.
-->

fables-tales committed:
Introduce "RipperTree" as first stitching layer

fables-tales committed:
continued refactoring

fables-tales committed:
Replace ruby_init with ruby_setup, move to op

fables-tales committed:
zomg

fables-tales committed:
implement better comment extraction

fables-tales committed:
fmt

fables-tales committed:
Comment tests work!

fables-tales committed:
fmt

fables-tales committed:
Various final fixes

sgrif committed:
Fix double free in `debug_inspect` (#231)

This was the cause of the segfault that earlier commits in this branch
tried to fix. They actually fixed it by not calling `debug_inspect`, the
way it was performing iteration was fine (other than the signature of
the callback missing an argument, which could fuck things up depending
on the ABI of the target).

The source of this bug is that `CString` is being constructed with a
pointer we don't own. `CString` is an owned string, which frees its
memory on drop. This results in a double free at best.

The reason this manifested as it did was because we're not just freeing
a pointer we don't own, we're freeing a pointer that was allocated by a
different allocator. Since we didn't build ruby with jemalloc, and we
aren't compiling our jemalloc with unprefixed symbols (we should totally
do that), jemalloc would end up corrupting its own book keeping when we
freed it.

This is also why the segfault disappeared if we used the system
allocator in Rust, or unprefixed symbols in jemalloc. When we do either
of those, it becomes "just" a double free, since the string is never
read again. Even the double free was unlikely to manifest since GC
almost certainly never runs again after Ripper finishes parsing.

sgrif committed:
Partial audit of unsafe code, refactor to reduce its usage (#232)

* Partial audit of unsafe code, refactor to reduce its usage

This is the result of a partial audit of all unsafe code in rubyfmt. I
found one double free, and many instances where unsafe was used more
than it was needed. In some cases this was caused by omitting extremely
cheap safety checks like UTF-8 validation, and in some cases it was
redundant code converting arrays and strings to Rust versions. I've
added UTF-8 checking where it's clearly not going to cause a performance
problem, commented where it's not clear, and moved array and string
conversion to a common function that's easier to audit.

* cargo fmt

fables-tales committed:
extract render queue to basequeue

fables-tales committed:
refactor targets

fables-tales committed:
finish interface extraction

fables-tales committed:
plz

sgrif committed:
make it compile

fables-tales committed:
extract everything to current target

fables-tales committed:
slightly better comment handling

fables-tales committed:
Make comments better

fables-tales committed:
omg make it work

fables-tales committed:
fmt

fables-tales committed:
add another large-ish ruby file

fables-tales committed:
blow cargo cache

fables-tales committed:
fix caching dirs

sgrif commented on librubyfmt/src/comment_block.rs:
```diff
@@ -13,10 +14,25 @@ impl CommentBlock {
+                let spaces = (0..indent_depth)
+                    .map(|_| " ".to_string())
+                    .collect::<Vec<String>>()
+                    .join("");
```
```suggestion
                let spaces = str::repeat(" ", indent_depth)
```

sgrif commented on librubyfmt/src/ruby_ops.rs:
```diff
@@ -0,0 +1,82 @@
+}
+
+#[derive(Clone, Copy, Debug)]
+pub struct RipperTree(VALUE);
```
Given that this just has a safe constructor and never asserts that the inner type is actually what is expected, I'm not entirely sure what value this brings

sgrif commented on librubyfmt/src/file_comments.rs:
```diff
@@ -1,51 +1,54 @@
-                }
-                res.comment_blocks.insert(line_number, l);
-            }
+        let should_push =
+            line_number == 1 || (last_line.is_some() && last_line.unwrap() == &(line_number - 1));
```
```suggestion
        let should_push = last_line.copied().unwrap_or(0) == line_number - 1
```

sgrif replied:
Or if my assumption that `last_line` is always `None` when `line_number == 1` is incorrect,
```suggestion
        let should_push =
            line_number == 1 || last_line.copied() == Some(line_number - 1);
```

sgrif commented on librubyfmt/src/file_comments.rs:
```diff
@@ -1,51 +1,54 @@
+        let should_push =
+            line_number == 1 || (last_line.is_some() && last_line.unwrap() == &(line_number - 1));
+        if should_push {
+            self.contiguous_starting_indices.push(line_number);
```
I need to give a higher level overview of what this is for, but my gut tells me this is sidestepping the abstraction we need or that what this is could be made more clear

sgrif commented on librubyfmt/src/file_comments.rs:
```diff
@@ -59,7 +62,11 @@ impl FileComments {
+            sled.push(
+                self.comment_blocks
+                    .remove(key)
+                    .unwrap_or_else(|| panic!("we tracked it: {} {:?}", key, self.comment_blocks)),
```
This panic is also making me think there's a better structure for this

sgrif commented on librubyfmt/src/lib.rs:
```diff
@@ -76,28 +76,30 @@ pub enum FormatError {
-        ruby::ruby_init();
+    let res = ruby_ops::setup_ruby();
+    if res.is_err() {
+        return InitStatus::ERROR as libc::c_int;
```
Should the C API provide a way to get a human readable description of the error?

```rust
static mut LAST_ERROR: Option<CString> = None;

// on this line
if let Err(e) = res {
    // If we think there could be multi-threading here we should use `lazy_static!` and a mutex
    unsafe { LAST_ERROR = Some(CString::new(e.to_string()).expect("don't actually expect in prod")) }
    return InitStatus::ERROR as libc::c_int;
}

// later
#[no_mangle]
pub extern "C" fn rubyfmt_last_error -> *const libc::c_char {
    match LAST_ERROR {
        Some(e) => e.as_ptr(),
        None => ptr::null(),
    }
}
```

fables-tales replied:
the header defines this: https://github.com/penelopezone/rubyfmt/blob/trunk/librubyfmt/include/rubyfmt.h#L4

sgrif commented on librubyfmt/src/parser_state.rs:
```diff
@@ -320,22 +312,21 @@ impl ParserState {
 
+            let spaces = self.spaces_after_last_newline;
+            debug!("spaces: {} comments: {:?}", spaces, self.comments_to_insert);
+            let bt = Backtrace::new();
+            debug!("{:?}", bt);
```
If we're at the point where we're printing backtraces into the debug logger, we should probably use the `tracing` library instead

fables-tales replied:
this was pretty one-off. I'll probably just delete

sgrif commented on librubyfmt/src/parser_state.rs:
```diff
@@ -487,6 +494,9 @@ impl ParserState {
+        // not be UTF-8
         unsafe {
             let s = str::from_utf8_unchecked(&data).to_string();
             s.trim().chars().any(|v| v == '\n')
```
```suggestion
            }
        s.trim().contains('\n')
```

(It won't let me include the closing brace that I'm moving in this suggestion comment sorry

fables-tales replied:
manually applied

sgrif commented on librubyfmt/src/parser_state.rs:
```diff
@@ -503,10 +513,7 @@ impl ParserState {
-                Some(x) => !x.is_newline(),
-                None => true,
-            };
+            let want_newline = !self.current_target_mut().last_token_is_a_newline();
```
```suggestion
            let want_newline = !self.current_target().last_token_is_a_newline();
```

sgrif commented on librubyfmt/src/parser_state.rs:
```diff
@@ -624,4 +624,24 @@ impl ParserState {
     }
+
+    fn current_target(&self) -> &dyn LineTokenTarget {
+        if self.breakable_entry_stack.is_empty() {
+            &self.render_queue
+        } else {
+            self.breakable_entry_stack
+                .last()
+                .expect("we checked it's not empty")
+        }
```
```suggestion
        self.breakable_entry_stack
            .last()
            .unwrap_or(&self.render_queue)
```

sgrif commented on librubyfmt/src/parser_state.rs:
```diff
@@ -624,4 +624,24 @@ impl ParserState {
+    }
+
+    fn current_target_mut(&mut self) -> &mut dyn LineTokenTarget {
+        if self.breakable_entry_stack.is_empty() {
+            &mut self.render_queue
+        } else {
+            self.breakable_entry_stack
+                .last_mut()
+                .expect("we checked it's not empty")
+        }
```
```suggestion
        self.breakable_entry_stack
            .last_mut()
            .unwrap_or(&mut self.render_queue)
```

sgrif commented on librubyfmt/src/render_targets.rs:
```diff
@@ -3,11 +3,58 @@ use crate::line_tokens::LineToken;
 use std::collections::HashSet;
 
+fn insert_at<T>(idx: usize, target: &mut Vec<T>, input: &mut Vec<T>) {
+    let drain = input.drain(..);
+    let mut idx = idx;
+    for item in drain {
+        target.insert(idx, item);
+        idx += 1;
+    }
```
```suggestion
    let mut tail = target.split_off(idx);
    input.append(input);
    input.append(&mut tail);
```

sgrif replied:
The code as written will perform as many as log(N) allocations and 2N memcpys. The new code will perform at most 2 and at minimum 1 allocations, and 2 memcpys

sgrif commented on librubyfmt/src/render_targets.rs:
```diff
@@ -3,11 +3,58 @@ use crate::line_tokens::LineToken;
+        self.tokens.push(lt)
+    }
+
+    fn insert_at(&mut self, idx: usize, tokens: &mut Vec<LineToken>) {
```
It looks like this is mostly called with a single element, should this just take that element instead of a vec?

sgrif commented on librubyfmt/src/render_targets.rs:
```diff
@@ -3,11 +3,58 @@ use crate::line_tokens::LineToken;
     SingleLine,
 }
 
+pub trait LineTokenTarget {
+    fn push(&mut self, lt: LineToken);
+    fn insert_at(&mut self, idx: usize, tokens: &mut Vec<LineToken>);
+    fn into_tokens(self, ct: ConvertType) -> Vec<LineToken>;
+    fn last_token_is_a_newline(&self) -> bool;
+    fn index_of_prev_hard_newline(&self) -> Option<usize>;
+}
```
The impl of every method on this trait is exactly the same (or could be). How about this instead?
```suggestion
pub trait LineTokenTarget {
    fn tokens(&self) -> &[LineToken];
    fn tokens_mut(&mut self) -> &mut Vec<LineToken>;
    
    fn push(&mut self, lt: LineToken) {
        self.tokens_mut().push(lt);
    }
    
    // etc
}
```

sgrif commented on librubyfmt/src/render_targets.rs:
```diff
@@ -47,6 +89,30 @@ impl BreakableEntry {
+    pub fn new(spaces: ColNumber, delims: BreakableDelims) -> Self {
+        BreakableEntry {
+            spaces,
+            tokens: vec![],
```
Seeing `vec![]` next to `HashSet::new()` feels weird to me

```suggestion
            tokens: Vec::new(),
```

fables-tales replied:
applied manually

sgrif commented on librubyfmt/src/ruby.rs:
```diff
@@ -81,29 +83,82 @@ extern "C" {
 }
 
 pub fn current_exception_as_rust_string() -> String {
+    let ruby_string = unsafe { eval_str("$!.inspect") }.expect("Error evaluating `$!.inspect`");
```
Rather than panicking, how about we return a string indicating that there was an error even trying to get the value out of Ruby. Or even better, is there a function in the Ruby C API that we can call instead of evaling?

sgrif commented on librubyfmt/src/ruby.rs:
```diff
@@ -81,29 +83,82 @@ extern "C" {
+
+// Safety: This function expects an initialized Ruby VM capable of evaling code
+pub unsafe fn eval_str(s: &str) -> Result<VALUE, ()> {
+    let rubyfmt_program_as_c = CString::new(s).expect("unexpected nul byte in Ruby code");
```
Can we return an error instead of panicking?

sgrif commented on librubyfmt/src/ruby.rs:
```diff
@@ -81,29 +83,82 @@ extern "C" {
-        let ptr = rubyfmt_rstring_ptr(res);
-        let length = rubyfmt_rstring_len(res);
-        String::from_raw_parts(ptr as _, length as _, length as _)
+        let inspect = rb_funcall(v, intern!("inspect"), 0, std::ptr::null() as *const VALUE);
```
```suggestion
        let inspect = rb_funcall(v, intern!("inspect"), 0, std::ptr::null());
```

sgrif commented on librubyfmt/src/ruby.rs:
```diff
@@ -81,29 +83,82 @@ extern "C" {
-        let length = rubyfmt_rstring_len(res);
-        String::from_raw_parts(ptr as _, length as _, length as _)
+        let inspect = rb_funcall(v, intern!("inspect"), 0, std::ptr::null() as *const VALUE);
+        let char_pointer = rb_string_value_cstr(&inspect) as *mut i8;
+        let cstr = CStr::from_ptr(char_pointer);
+        let s = cstr.to_str().expect("it's utf8");
+        debug!("{}", s);
```
```suggestion
        debug!("{}", unsafe { ruby_string_to_str(inspect) });
```

sgrif commented on librubyfmt/src/ruby.rs:
```diff
@@ -50,13 +51,13 @@ pub const Qnil: VALUE = VALUE(8);
     pub fn rb_utf8_str_new(_: *const libc::c_char, _: libc::c_long) -> VALUE;
     pub fn rb_str_new_cstr(_: *const libc::c_char) -> VALUE;
-    pub fn rb_string_value_cstr(_: VALUE) -> *const libc::c_char;
+    pub fn rb_string_value_cstr(_: *const VALUE) -> *const libc::c_char;
```
I don't see a corresponding change to the definition. Are you sure this is correct?

fables-tales replied:
https://github.com/ruby/ruby/blob/7bafc9f433223d82aac8e8974b9bfdbdc34ca4df/include/ruby/internal/core/rstring.h#L92

sgrif commented on librubyfmt/src/ruby.rs:
```diff
@@ -81,29 +83,82 @@ extern "C" {
-        } else {
-            Ok(v)
+            let s = current_exception_as_rust_string();
+            panic!("blew us: {}", s);
```
Are you sure panicking if we failed to perform a debug inspection is the right call?

sgrif commented on librubyfmt/src/ruby.rs:
```diff
@@ -81,29 +83,82 @@ extern "C" {
 }
+
+pub fn raise(s: &str) {
+    let cstr = CString::new(s).expect("it's not null");
```
"it's not null" doesn't accurately describe this error condition. Can we just handle a nul byte being contained in the string gracefully or at least give a useful error message?

sgrif reviewed (approved):
I don't see any obvious UB. I would still like to improve the structure of a lot of this and/or attempt to make it safe in the future. I also had a few comments about style. I really think you should try to reduce the number of places this code base uses panics. But overall I don't see anything too heinous

sgrif committed:
Refactor `FileComments` to better demonstrate intent

I'm not particularly happy with how this came out, but I do think it's a
clear improvement over what was there before. Previously this has three
obviously interacting fields, with a lot of subtle, undocumented,
unmaintained invariants (or at least in the case of `lowest_key` it was
unmaintained). I do believe that this structure at least better
represents intent.

`CommentBlock` now keeps track of what lines it's on, and the "start of
file sled" (what even is that?!) is kept separately from the rest of the
comments. This reduces (but does not eliminates) the amount of state
that this type has to keep, and makes the intent clearer both in the
type itself, and its state at any single point. At absolute minimum, I
have at least removed all panics from this type's implementation.

That said, this interface has a ton of pitfalls. I really dislike how
much it needs to mutate itself. While developing this patch, it became
clear to me why this is the case. The parser that we've written only
retains the line number and the comment on that line. This means we have
no way to determine if two comments are part of a contiguous block or
not, and it seems we leave the formatter to determine whether that's the
case. We do treat the first comment block specially though, and assume
it's a contiguous block (though I'm not entirely clear why yet).

This is a problem for a number of reasons. First of all, it means that
we will just delete comments if there are more than one on a given line,
potentially breaking comments on method arguments that are significant
to rdoc.

This also misbehaves if the first non-comment line has a trailing
comment (most likely for rdoc, but could also potentially happen if the
first comment is for the file itself, and the trailing comment is a
poorly placed, very short comment for the item it's on)

This type very clearly wants to be `Map<Range<LineNumber>, Vec<String>>`
or even better `Vec<Span, String>` since the formatter should be able to
choose/change where newlines are inserted, and really shouldn't be
coupled to line numbers in the source. I'm sure Ripper *must* give us a
way to determine if non-comment tokens appear between two comments, but
I do not know enough about that layer of the code to fix it in this
patch. The fact that `FileComments` is constructed outside of the main
deserializer is surprising, and I would hope that we can perhaps at
least handle this in a `Deserialize` impl`.

But short of what I'd consider to be a more complete fix, I have at
least tried to make the invariants of this type better upheld, and those
that aren't enforced are now documented (mostly in `push_comment`, which
is still extremely subtle)

I don't think `push_comment` is the right interface for this type.
However, since it was marked as public, I've assumed that it is an
interface intended to be externally consumed and thus needs to be
maintained. If that's not the case, I think we should replace it with
something that only operates when we have a full view of the data, such
as `FromIterator`. In that case we can either manually sort the data
before consuming it, or at least include a debug assertion that it's
already sorted.

Speaking of debug assertions, one thing that stood out to me while
writing this is that the test suite is run in release mode. I'd like to
learn more about why that is.  Previous iterations of this patch
included debug assertions for the invariants that were implied but not
upheld. While none of these landed in the final patch, if they did I
feel like they should be run as part of the test suite.

fables-tales committed:
just a little bit of cleanup

fables-tales committed:
Update librubyfmt/src/comment_block.rs

Co-authored-by: Siân Griffin <sean@seantheprogrammer.com>

fables-tales committed:
Update librubyfmt/src/ruby.rs

Co-authored-by: Siân Griffin <sean@seantheprogrammer.com>

fables-tales committed:
Update librubyfmt/src/ruby.rs

Co-authored-by: Siân Griffin <sean@seantheprogrammer.com>

fables-tales committed:
Update librubyfmt/src/file_comments.rs

Co-authored-by: Siân Griffin <sean@seantheprogrammer.com>

fables-tales committed:
suggestion

fables-tales committed:
Update librubyfmt/src/parser_state.rs

Co-authored-by: Siân Griffin <sean@seantheprogrammer.com>

fables-tales committed:
some stuff

fables-tales committed:
Update librubyfmt/src/render_targets.rs

Co-authored-by: Siân Griffin <sean@seantheprogrammer.com>

fables-tales committed:
fix ci

fables-tales committed:
less debug

fables-tales committed:
Update librubyfmt/src/parser_state.rs

Co-authored-by: Siân Griffin <sean@seantheprogrammer.com>

fables-tales committed:
better blowup string

fables-tales committed:
Merge branch 'refactor-ruby-interface' of github.com:penelopezone/rubyfmt into refactor-ruby-interface

fables-tales committed:
Revert "Update librubyfmt/src/parser_state.rs"

This reverts commit 6004a0dd13057d3df90dd20d5e8e589e04b1f693.

fables-tales committed:
remove unused line

fables-tales committed:
fix
